test/Microsoft.Azure.Devices.Edge.Test.Common/Linux/EdgeDaemon.cs

```diff
@@ -0,0 +1,209 @@
+                {
+                    string[] output = await Process.RunAsync("bash", $"-c \"{string.Join("; ", commands)}\"", token);
+                    Log.Verbose(string.Join("\n", output));
+                },
+                message,
+                properties);
+
+            string hostname = (await File.ReadAllTextAsync("/proc/sys/kernel/hostname", token)).Trim();
+            IotHubConnectionStringBuilder builder = IotHubConnectionStringBuilder.Create(deviceConnectionString);
+
+            await Profiler.Run(
+                async () =>
+                {
+                    await this.InternalStopAsync(token);
+
+                    const string YamlPath = "/etc/iotedge/config.yaml";
+                    string text = await File.ReadAllTextAsync(YamlPath, token);
+
+                    var doc = new YamlDocument(text);
+                    doc.ReplaceOrAdd("provisioning.device_connection_string", deviceConnectionString);
+                    doc.ReplaceOrAdd("hostname", hostname);
+
+                    string result = doc.ToString();
+
+                    FileAttributes attr = 0;
+                    if (File.Exists(YamlPath))
+                    {
+                        attr = File.GetAttributes(YamlPath);
+                        File.SetAttributes(YamlPath, attr & ~FileAttributes.ReadOnly);
+                    }
+
+                    await File.WriteAllTextAsync(YamlPath, result, token);
+
+                    if (attr != 0)
+                    {
+                        File.SetAttributes(YamlPath, attr);
+                    }
+
+                    await this.InternalStartAsync(token);
+                },
+                "Configured edge daemon for device '{Device}' registered as '{Id}'",
+                hostname,
+                builder.DeviceId);
+        }
+
+        public Task StartAsync(CancellationToken token)
+        {
+            return Profiler.Run(
+                () => this.InternalStartAsync(token),
+                "Started edge daemon");
+        }
+
+        async Task InternalStartAsync(CancellationToken token)
+        {
+            string[] output = await Process.RunAsync("systemctl", "start iotedge", token);
+            Log.Verbose(string.Join("\n", output));
+            await this.WaitForStatusAsync(ServiceControllerStatus.Running, token);
+        }
+
+        public Task StopAsync(CancellationToken token)
+        {
+            return Profiler.Run(
+                () => this.InternalStopAsync(token),
+                "Stopped edge daemon");
+        }
+
+        async Task InternalStopAsync(CancellationToken token)
+        {
+            string[] output = await Process.RunAsync("systemctl", "stop iotedge", token);
+            Log.Verbose(string.Join("\n", output));
+            await this.WaitForStatusAsync(ServiceControllerStatus.Stopped, token);
+        }
+
+        public async Task UninstallAsync(CancellationToken token)
+        {
+            try
```

Comment:
should we check whether iotedge service installed first?

Reply:
This is the way we do it in quickstart as well. We could check first, but that would be one more call, and effectively the same result. It would allow us to remove the try/catch, which is nice. But then we'd want to also get rid of the try/catch for stopping the service, but we can't because it's not atomic. So I think it's just easier to keep it the way it is.

Reply:
I think checking service installed is to ensure valid input for this method; and this check should print a message if it is not installed.  your call.

Reply:
I looked into this a little more...it's certainly doable but it just seems like a slightly different approach to solving the same problem--it's not clearly better or worse. The function already expects that iotedged might not be installed, and handles things correctly if it isn't. I'd like to just keep what I have for now.